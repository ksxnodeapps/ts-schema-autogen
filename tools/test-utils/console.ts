import { ConsoleInstance } from 'simple-fake-console'
import { Console } from '@ts-schema-autogen/types'

export class FakeConsole implements Console.Mod {
  public readonly core = new ConsoleInstance()
  public readonly info = jest.fn(this.core.info)
  public readonly error = jest.fn(this.core.error)
}
