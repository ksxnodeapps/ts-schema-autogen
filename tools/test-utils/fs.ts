import { StringPathFileSystem } from 'simple-fake-fs'
import { FSX } from '@ts-schema-autogen/types'

export class FakeFileSystem extends StringPathFileSystem implements FSX.Mod {
  private mkfn<Xs extends any[], Y> (fn: (...xs: Xs) => Y) {
    return jest.fn(async (...xs: Xs) => fn.apply(this, xs))
  }

  public readonly stat = this.mkfn(this.statSync)
  public readonly readdir = this.mkfn(this.readdirSync)
  public readonly readFile = this.mkfn(this.readFileSync)
  public readonly outputFile = this.mkfn(this.outputFileSync)
  public readonly writeFile = this.mkfn(this.writeFileSync)
  public readonly remove = jest.fn()
}
