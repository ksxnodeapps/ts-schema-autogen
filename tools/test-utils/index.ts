export * from './console'
export * from './fs'
export * from './path'
export * from './print-result'
export * from './process'
export * from './typescript-json-schema'
