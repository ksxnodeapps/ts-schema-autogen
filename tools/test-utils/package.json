{
  "name": "@tools/test-utils",
  "version": "0.0.0",
  "private": true,
  "dependencies": {
    "simple-fake-fs": "^0.0.8",
    "simple-fake-console": "^0.1.6",
    "simple-fake-path": "^0.0.2",
    "@ts-schema-autogen/status": "file:../../packages/status",
    "@ts-schema-autogen/types": "file:../../packages/types"
  }
}
