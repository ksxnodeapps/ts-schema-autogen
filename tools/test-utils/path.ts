import { Path } from '@ts-schema-autogen/types'

export class FakePath implements Path.Mod {
  constructor (private readonly cwd: string) {}

  public readonly join = (...paths: string[]): string => {
    if (!paths.length) return '.'
    const [head, ...rest] = paths
    const tail = this.join(...rest)
    if (head === '.') return tail
    if (tail === '.') return head
    if (tail.startsWith('../')) return this.join(this.dirname(head), tail.slice('../'.length))
    const left = head.endsWith('/') ? head.slice(0, -1) : head
    const right = tail.startsWith('/') ? tail.slice(1) : tail
    return left + '/' + right
  }

  public readonly resolve = (...paths: string[]): string => {
    if (!paths.length) return '.'
    const [head, ...rest] = paths
    const tail = this.resolve(...rest)
    if (tail.startsWith('/')) return tail
    return this.join(this.cwd, head, tail)
  }

  public readonly dirname = (path: string): string => {
    if (path === '' || path === '.' || path === '/') return path
    const segments = path.split('/').slice(0, -1)
    if (!segments.length) return '.'
    return this.join(...segments)
  }

  public readonly basename = (path: string): string => {
    const segments = path.split('/')
    return segments[segments.length - 1]
  }
}
