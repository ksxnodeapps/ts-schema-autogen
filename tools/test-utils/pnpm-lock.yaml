dependencies:
  '@ts-schema-autogen/types': 'link:../../packages/types'
  simple-fake-fs: 0.0.4
lockfileVersion: 5.1
packages:
  /@types/node/13.5.0:
    dev: false
    resolution:
      integrity: sha512-Onhn+z72D2O2Pb2ql2xukJ55rglumsVo1H6Fmyi8mlU9SvKdBk/pUSUAiBY/d9bAOF7VVWajX3sths/+g6ZiAQ==
  /simple-fake-fs/0.0.4:
    dependencies:
      '@types/node': 13.5.0
      tslib: 1.10.0
    dev: false
    resolution:
      integrity: sha512-GUwEIYXI6oSJdMZkP1hK3x084z6HX6uyHMrs3erJXEzdyMIrJ7HJEjlbtMrs6zsELvgx0wr7Y9zTK61+igHC8A==
  /tslib/1.10.0:
    dev: false
    resolution:
      integrity: sha512-qOebF53frne81cf0S9B41ByenJ3/IuH8yJKngAX35CmiZySA0khhkovshKK+jGCaMnVomla7gVlIcc3EvKPbTQ==
specifiers:
  '@ts-schema-autogen/types': ../../packages/types
  simple-fake-fs: ^0.0.4
