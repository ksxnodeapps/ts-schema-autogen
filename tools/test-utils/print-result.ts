import { inspect } from 'util'
import { Success, Failure } from '@ts-schema-autogen/status'

export function printFailure (result: Failure<unknown>): string {
  let text = '\n'
  result.print((...args) => text += args.join(' ') + '\n')
  return text
}

export const printResult = (result: Success<unknown> | Failure<unknown>): string =>
  result.code ? printFailure(result) : inspect(result)
