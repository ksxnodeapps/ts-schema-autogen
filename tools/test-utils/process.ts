import { Process } from '@ts-schema-autogen/types'

export class FakeProcess implements Process.Mod {
  constructor (private readonly workDir: string) {}
  public readonly cwd = jest.fn(() => this.workDir)
}
