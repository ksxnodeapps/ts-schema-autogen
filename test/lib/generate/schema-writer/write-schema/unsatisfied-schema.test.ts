import fsTree from '../../../../fixtures/fs-tree/unsatisfied-schema'
import TJS from '../../../../fixtures/tjs'
import { FakeFileSystem, FakePath, printResult } from '@tools/test-utils'

import {
  Status,
  UnsatisfiedSchema,
  MultipleFailures,
  SchemaWriter,
  createJsonConfigParser
} from '@ts-schema-autogen/lib'

async function setup (configPaths: readonly string[]) {
  const fsx = new FakeFileSystem('/', fsTree)
  const path = new FakePath()
  const tjs = new TJS()
  const schemaWriter = new SchemaWriter({
    fsx,
    path,
    tjs,
    parsers: [
      await createJsonConfigParser('JSON Parser')
    ]
  })
  const result = await schemaWriter.writeSchemas(configPaths)
  return { fsx, path, tjs, result }
}

describe('single failure', () => {
  const configPaths = [
    'multiple-errors/.schema.autogen.json'
  ]

  it('returns result matching snapshot', async () => {
    const { result } = await setup(configPaths)
    expect(result).toMatchSnapshot()
  })

  it('returns an UnsatisfiedSchema', async () => {
    const { result } = await setup(configPaths)
    expect(result).toBeInstanceOf(UnsatisfiedSchema)
  })

  it('returns result containing expected properties', async () => {
    const { result } = await setup(configPaths)
    expect(result).toMatchObject({
      code: Status.UnsatisfiedSchema,
      error: expect.any(Array)
    })
  })

  it('does not call outputFile', async () => {
    const { fsx } = await setup(configPaths)
    expect(fsx.outputFile).not.toBeCalled()
  })

  it('error messages', async () => {
    const { result } = await setup(configPaths)
    expect(printResult(result)).toMatchSnapshot()
  })

  it('status code', async () => {
    const { result } = await setup(configPaths)
    expect(result.getStatusCode()).toBe(Status.UnsatisfiedSchema)
  })
})

describe('multiple failures', () => {
  const configPaths = [
    'string-instead-of-object/.schema.autogen.json',
    'missing-instruction/.schema.autogen.json',
    'invalid-instruction/.schema.autogen.json',
    'multiple-errors/.schema.autogen.json'
  ]

  it('returns result matching snapshot', async () => {
    const { result } = await setup(configPaths)
    expect(result).toMatchSnapshot()
  })

  it('returns an MultipleFailures', async () => {
    const { result } = await setup(configPaths)
    expect(result).toBeInstanceOf(MultipleFailures)
  })

  it('returns result containing expected properties', async () => {
    const { result } = await setup(configPaths)
    expect(result).toMatchObject({
      code: Status.MultipleFailures,
      error: expect.any(Array)
    })
  })

  it('does not call outputFile', async () => {
    const { fsx } = await setup(configPaths)
    expect(fsx.outputFile).not.toBeCalled()
  })

  it('error messages', async () => {
    const { result } = await setup(configPaths)
    expect(printResult(result)).toMatchSnapshot()
  })

  it('status code', async () => {
    const { result } = await setup(configPaths)
    expect(result.getStatusCode()).toBe(Status.UnsatisfiedSchema)
  })
})
