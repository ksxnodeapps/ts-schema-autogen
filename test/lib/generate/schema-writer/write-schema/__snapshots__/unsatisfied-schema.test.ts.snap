// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiple failures error messages 1`] = `
"
MultipleFailures
    UnsatisfiedSchema
        path: string-instead-of-object/.schema.autogen.json
        error:
            <data> should be object
    UnsatisfiedSchema
        path: missing-instruction/.schema.autogen.json
        error:
            <data> should have required property 'instruction'
    UnsatisfiedSchema
        path: invalid-instruction/.schema.autogen.json
        error:
            <data>.instruction should be object
            <data>.instruction should be object
            <data>.instruction should match some schema in anyOf
    UnsatisfiedSchema
        path: multiple-errors/.schema.autogen.json
        error:
            <data>.instruction.output should be null
            <data>.instruction.symbol should be null
            <data>.instruction.compilerOptions.allowJs should be boolean
            <data>.instruction.compilerOptions.checkJs should be boolean
            <data>.instruction.compilerOptions.esModuleInterop should be boolean
            <data>.instruction.schemaSettings.excludePrivate should be boolean
            <data>.instruction.schemaSettings.defaultNumberType should be equal to one of the allowed values
                allowed values: [ 'integer', 'number' ]
            <data>.instruction.input[1] should be string
            <data>.instruction.input[2] should be string
            <data>.instruction.input should be string
            <data>.instruction.input should match some schema in anyOf
            <data>.instruction.compilerOptions.allowJs should be boolean
            <data>.instruction.compilerOptions.checkJs should be boolean
            <data>.instruction.compilerOptions.esModuleInterop should be boolean
            <data>.instruction.schemaSettings.excludePrivate should be boolean
            <data>.instruction.schemaSettings.defaultNumberType should be equal to one of the allowed values
                allowed values: [ 'integer', 'number' ]
            <data>.instruction.input[1] should be string
            <data>.instruction.input[2] should be string
            <data>.instruction.input should be string
            <data>.instruction.input should match some schema in anyOf
            <data>.instruction.output should be object
            <data>.instruction.output[2].filename should be string
            <data>.instruction.output[2] should be string
            <data>.instruction.output[2] should match some schema in anyOf
            <data>.instruction.output[3] should have required property 'filename'
            <data>.instruction.output[3] should be string
            <data>.instruction.output[3] should match some schema in anyOf
            <data>.instruction.output should be string
            <data>.instruction.output should match some schema in anyOf
            <data>.instruction should match some schema in anyOf
"
`;

exports[`multiple failures returns result matching snapshot 1`] = `
MultipleFailures {
  "code": 2,
  "error": Array [
    UnsatisfiedSchema {
      "code": 12,
      "error": Array [
        Object {
          "dataPath": "",
          "keyword": "type",
          "message": "should be object",
          "params": Object {
            "type": "object",
          },
          "schemaPath": "#/type",
        },
      ],
      "path": "string-instead-of-object/.schema.autogen.json",
      "value": undefined,
    },
    UnsatisfiedSchema {
      "code": 12,
      "error": Array [
        Object {
          "dataPath": "",
          "keyword": "required",
          "message": "should have required property 'instruction'",
          "params": Object {
            "missingProperty": "instruction",
          },
          "schemaPath": "#/required",
        },
      ],
      "path": "missing-instruction/.schema.autogen.json",
      "value": undefined,
    },
    UnsatisfiedSchema {
      "code": 12,
      "error": Array [
        Object {
          "dataPath": ".instruction",
          "keyword": "type",
          "message": "should be object",
          "params": Object {
            "type": "object",
          },
          "schemaPath": "#/type",
        },
        Object {
          "dataPath": ".instruction",
          "keyword": "type",
          "message": "should be object",
          "params": Object {
            "type": "object",
          },
          "schemaPath": "#/type",
        },
        Object {
          "dataPath": ".instruction",
          "keyword": "anyOf",
          "message": "should match some schema in anyOf",
          "params": Object {},
          "schemaPath": "#/properties/instruction/anyOf",
        },
      ],
      "path": "invalid-instruction/.schema.autogen.json",
      "value": undefined,
    },
    UnsatisfiedSchema {
      "code": 12,
      "error": Array [
        Object {
          "dataPath": ".instruction.output",
          "keyword": "type",
          "message": "should be null",
          "params": Object {
            "type": "null",
          },
          "schemaPath": "#/properties/output/type",
        },
        Object {
          "dataPath": ".instruction.symbol",
          "keyword": "type",
          "message": "should be null",
          "params": Object {
            "type": "null",
          },
          "schemaPath": "#/properties/symbol/type",
        },
        Object {
          "dataPath": ".instruction.compilerOptions.allowJs",
          "keyword": "type",
          "message": "should be boolean",
          "params": Object {
            "type": "boolean",
          },
          "schemaPath": "#/properties/compilerOptions/properties/allowJs/type",
        },
        Object {
          "dataPath": ".instruction.compilerOptions.checkJs",
          "keyword": "type",
          "message": "should be boolean",
          "params": Object {
            "type": "boolean",
          },
          "schemaPath": "#/properties/compilerOptions/properties/checkJs/type",
        },
        Object {
          "dataPath": ".instruction.compilerOptions.esModuleInterop",
          "keyword": "type",
          "message": "should be boolean",
          "params": Object {
            "type": "boolean",
          },
          "schemaPath": "#/properties/compilerOptions/properties/esModuleInterop/type",
        },
        Object {
          "dataPath": ".instruction.schemaSettings.excludePrivate",
          "keyword": "type",
          "message": "should be boolean",
          "params": Object {
            "type": "boolean",
          },
          "schemaPath": "#/definitions/Partial<Args>/properties/excludePrivate/type",
        },
        Object {
          "dataPath": ".instruction.schemaSettings.defaultNumberType",
          "keyword": "enum",
          "message": "should be equal to one of the allowed values",
          "params": Object {
            "allowedValues": Array [
              "integer",
              "number",
            ],
          },
          "schemaPath": "#/definitions/Partial<Args>/properties/defaultNumberType/enum",
        },
        Object {
          "dataPath": ".instruction.input[1]",
          "keyword": "type",
          "message": "should be string",
          "params": Object {
            "type": "string",
          },
          "schemaPath": "#/properties/input/anyOf/0/items/type",
        },
        Object {
          "dataPath": ".instruction.input[2]",
          "keyword": "type",
          "message": "should be string",
          "params": Object {
            "type": "string",
          },
          "schemaPath": "#/properties/input/anyOf/0/items/type",
        },
        Object {
          "dataPath": ".instruction.input",
          "keyword": "type",
          "message": "should be string",
          "params": Object {
            "type": "string",
          },
          "schemaPath": "#/properties/input/anyOf/1/type",
        },
        Object {
          "dataPath": ".instruction.input",
          "keyword": "anyOf",
          "message": "should match some schema in anyOf",
          "params": Object {},
          "schemaPath": "#/properties/input/anyOf",
        },
        Object {
          "dataPath": ".instruction.compilerOptions.allowJs",
          "keyword": "type",
          "message": "should be boolean",
          "params": Object {
            "type": "boolean",
          },
          "schemaPath": "#/properties/compilerOptions/properties/allowJs/type",
        },
        Object {
          "dataPath": ".instruction.compilerOptions.checkJs",
          "keyword": "type",
          "message": "should be boolean",
          "params": Object {
            "type": "boolean",
          },
          "schemaPath": "#/properties/compilerOptions/properties/checkJs/type",
        },
        Object {
          "dataPath": ".instruction.compilerOptions.esModuleInterop",
          "keyword": "type",
          "message": "should be boolean",
          "params": Object {
            "type": "boolean",
          },
          "schemaPath": "#/properties/compilerOptions/properties/esModuleInterop/type",
        },
        Object {
          "dataPath": ".instruction.schemaSettings.excludePrivate",
          "keyword": "type",
          "message": "should be boolean",
          "params": Object {
            "type": "boolean",
          },
          "schemaPath": "#/definitions/Partial<Args>/properties/excludePrivate/type",
        },
        Object {
          "dataPath": ".instruction.schemaSettings.defaultNumberType",
          "keyword": "enum",
          "message": "should be equal to one of the allowed values",
          "params": Object {
            "allowedValues": Array [
              "integer",
              "number",
            ],
          },
          "schemaPath": "#/definitions/Partial<Args>/properties/defaultNumberType/enum",
        },
        Object {
          "dataPath": ".instruction.input[1]",
          "keyword": "type",
          "message": "should be string",
          "params": Object {
            "type": "string",
          },
          "schemaPath": "#/properties/input/anyOf/0/items/type",
        },
        Object {
          "dataPath": ".instruction.input[2]",
          "keyword": "type",
          "message": "should be string",
          "params": Object {
            "type": "string",
          },
          "schemaPath": "#/properties/input/anyOf/0/items/type",
        },
        Object {
          "dataPath": ".instruction.input",
          "keyword": "type",
          "message": "should be string",
          "params": Object {
            "type": "string",
          },
          "schemaPath": "#/properties/input/anyOf/1/type",
        },
        Object {
          "dataPath": ".instruction.input",
          "keyword": "anyOf",
          "message": "should match some schema in anyOf",
          "params": Object {},
          "schemaPath": "#/properties/input/anyOf",
        },
        Object {
          "dataPath": ".instruction.output",
          "keyword": "type",
          "message": "should be object",
          "params": Object {
            "type": "object",
          },
          "schemaPath": "#/definitions/OutputDescriptor/type",
        },
        Object {
          "dataPath": ".instruction.output[2].filename",
          "keyword": "type",
          "message": "should be string",
          "params": Object {
            "type": "string",
          },
          "schemaPath": "#/definitions/OutputDescriptor/properties/filename/type",
        },
        Object {
          "dataPath": ".instruction.output[2]",
          "keyword": "type",
          "message": "should be string",
          "params": Object {
            "type": "string",
          },
          "schemaPath": "#/properties/output/anyOf/1/items/anyOf/1/type",
        },
        Object {
          "dataPath": ".instruction.output[2]",
          "keyword": "anyOf",
          "message": "should match some schema in anyOf",
          "params": Object {},
          "schemaPath": "#/properties/output/anyOf/1/items/anyOf",
        },
        Object {
          "dataPath": ".instruction.output[3]",
          "keyword": "required",
          "message": "should have required property 'filename'",
          "params": Object {
            "missingProperty": "filename",
          },
          "schemaPath": "#/definitions/OutputDescriptor/required",
        },
        Object {
          "dataPath": ".instruction.output[3]",
          "keyword": "type",
          "message": "should be string",
          "params": Object {
            "type": "string",
          },
          "schemaPath": "#/properties/output/anyOf/1/items/anyOf/1/type",
        },
        Object {
          "dataPath": ".instruction.output[3]",
          "keyword": "anyOf",
          "message": "should match some schema in anyOf",
          "params": Object {},
          "schemaPath": "#/properties/output/anyOf/1/items/anyOf",
        },
        Object {
          "dataPath": ".instruction.output",
          "keyword": "type",
          "message": "should be string",
          "params": Object {
            "type": "string",
          },
          "schemaPath": "#/properties/output/anyOf/2/type",
        },
        Object {
          "dataPath": ".instruction.output",
          "keyword": "anyOf",
          "message": "should match some schema in anyOf",
          "params": Object {},
          "schemaPath": "#/properties/output/anyOf",
        },
        Object {
          "dataPath": ".instruction",
          "keyword": "anyOf",
          "message": "should match some schema in anyOf",
          "params": Object {},
          "schemaPath": "#/properties/instruction/anyOf",
        },
      ],
      "path": "multiple-errors/.schema.autogen.json",
      "value": undefined,
    },
  ],
  "value": undefined,
}
`;

exports[`single failure error messages 1`] = `
"
UnsatisfiedSchema
    path: multiple-errors/.schema.autogen.json
    error:
        <data>.instruction.output should be null
        <data>.instruction.symbol should be null
        <data>.instruction.compilerOptions.allowJs should be boolean
        <data>.instruction.compilerOptions.checkJs should be boolean
        <data>.instruction.compilerOptions.esModuleInterop should be boolean
        <data>.instruction.schemaSettings.excludePrivate should be boolean
        <data>.instruction.schemaSettings.defaultNumberType should be equal to one of the allowed values
            allowed values: [ 'integer', 'number' ]
        <data>.instruction.input[1] should be string
        <data>.instruction.input[2] should be string
        <data>.instruction.input should be string
        <data>.instruction.input should match some schema in anyOf
        <data>.instruction.compilerOptions.allowJs should be boolean
        <data>.instruction.compilerOptions.checkJs should be boolean
        <data>.instruction.compilerOptions.esModuleInterop should be boolean
        <data>.instruction.schemaSettings.excludePrivate should be boolean
        <data>.instruction.schemaSettings.defaultNumberType should be equal to one of the allowed values
            allowed values: [ 'integer', 'number' ]
        <data>.instruction.input[1] should be string
        <data>.instruction.input[2] should be string
        <data>.instruction.input should be string
        <data>.instruction.input should match some schema in anyOf
        <data>.instruction.output should be object
        <data>.instruction.output[2].filename should be string
        <data>.instruction.output[2] should be string
        <data>.instruction.output[2] should match some schema in anyOf
        <data>.instruction.output[3] should have required property 'filename'
        <data>.instruction.output[3] should be string
        <data>.instruction.output[3] should match some schema in anyOf
        <data>.instruction.output should be string
        <data>.instruction.output should match some schema in anyOf
        <data>.instruction should match some schema in anyOf
"
`;

exports[`single failure returns result matching snapshot 1`] = `
UnsatisfiedSchema {
  "code": 12,
  "error": Array [
    Object {
      "dataPath": ".instruction.output",
      "keyword": "type",
      "message": "should be null",
      "params": Object {
        "type": "null",
      },
      "schemaPath": "#/properties/output/type",
    },
    Object {
      "dataPath": ".instruction.symbol",
      "keyword": "type",
      "message": "should be null",
      "params": Object {
        "type": "null",
      },
      "schemaPath": "#/properties/symbol/type",
    },
    Object {
      "dataPath": ".instruction.compilerOptions.allowJs",
      "keyword": "type",
      "message": "should be boolean",
      "params": Object {
        "type": "boolean",
      },
      "schemaPath": "#/properties/compilerOptions/properties/allowJs/type",
    },
    Object {
      "dataPath": ".instruction.compilerOptions.checkJs",
      "keyword": "type",
      "message": "should be boolean",
      "params": Object {
        "type": "boolean",
      },
      "schemaPath": "#/properties/compilerOptions/properties/checkJs/type",
    },
    Object {
      "dataPath": ".instruction.compilerOptions.esModuleInterop",
      "keyword": "type",
      "message": "should be boolean",
      "params": Object {
        "type": "boolean",
      },
      "schemaPath": "#/properties/compilerOptions/properties/esModuleInterop/type",
    },
    Object {
      "dataPath": ".instruction.schemaSettings.excludePrivate",
      "keyword": "type",
      "message": "should be boolean",
      "params": Object {
        "type": "boolean",
      },
      "schemaPath": "#/definitions/Partial<Args>/properties/excludePrivate/type",
    },
    Object {
      "dataPath": ".instruction.schemaSettings.defaultNumberType",
      "keyword": "enum",
      "message": "should be equal to one of the allowed values",
      "params": Object {
        "allowedValues": Array [
          "integer",
          "number",
        ],
      },
      "schemaPath": "#/definitions/Partial<Args>/properties/defaultNumberType/enum",
    },
    Object {
      "dataPath": ".instruction.input[1]",
      "keyword": "type",
      "message": "should be string",
      "params": Object {
        "type": "string",
      },
      "schemaPath": "#/properties/input/anyOf/0/items/type",
    },
    Object {
      "dataPath": ".instruction.input[2]",
      "keyword": "type",
      "message": "should be string",
      "params": Object {
        "type": "string",
      },
      "schemaPath": "#/properties/input/anyOf/0/items/type",
    },
    Object {
      "dataPath": ".instruction.input",
      "keyword": "type",
      "message": "should be string",
      "params": Object {
        "type": "string",
      },
      "schemaPath": "#/properties/input/anyOf/1/type",
    },
    Object {
      "dataPath": ".instruction.input",
      "keyword": "anyOf",
      "message": "should match some schema in anyOf",
      "params": Object {},
      "schemaPath": "#/properties/input/anyOf",
    },
    Object {
      "dataPath": ".instruction.compilerOptions.allowJs",
      "keyword": "type",
      "message": "should be boolean",
      "params": Object {
        "type": "boolean",
      },
      "schemaPath": "#/properties/compilerOptions/properties/allowJs/type",
    },
    Object {
      "dataPath": ".instruction.compilerOptions.checkJs",
      "keyword": "type",
      "message": "should be boolean",
      "params": Object {
        "type": "boolean",
      },
      "schemaPath": "#/properties/compilerOptions/properties/checkJs/type",
    },
    Object {
      "dataPath": ".instruction.compilerOptions.esModuleInterop",
      "keyword": "type",
      "message": "should be boolean",
      "params": Object {
        "type": "boolean",
      },
      "schemaPath": "#/properties/compilerOptions/properties/esModuleInterop/type",
    },
    Object {
      "dataPath": ".instruction.schemaSettings.excludePrivate",
      "keyword": "type",
      "message": "should be boolean",
      "params": Object {
        "type": "boolean",
      },
      "schemaPath": "#/definitions/Partial<Args>/properties/excludePrivate/type",
    },
    Object {
      "dataPath": ".instruction.schemaSettings.defaultNumberType",
      "keyword": "enum",
      "message": "should be equal to one of the allowed values",
      "params": Object {
        "allowedValues": Array [
          "integer",
          "number",
        ],
      },
      "schemaPath": "#/definitions/Partial<Args>/properties/defaultNumberType/enum",
    },
    Object {
      "dataPath": ".instruction.input[1]",
      "keyword": "type",
      "message": "should be string",
      "params": Object {
        "type": "string",
      },
      "schemaPath": "#/properties/input/anyOf/0/items/type",
    },
    Object {
      "dataPath": ".instruction.input[2]",
      "keyword": "type",
      "message": "should be string",
      "params": Object {
        "type": "string",
      },
      "schemaPath": "#/properties/input/anyOf/0/items/type",
    },
    Object {
      "dataPath": ".instruction.input",
      "keyword": "type",
      "message": "should be string",
      "params": Object {
        "type": "string",
      },
      "schemaPath": "#/properties/input/anyOf/1/type",
    },
    Object {
      "dataPath": ".instruction.input",
      "keyword": "anyOf",
      "message": "should match some schema in anyOf",
      "params": Object {},
      "schemaPath": "#/properties/input/anyOf",
    },
    Object {
      "dataPath": ".instruction.output",
      "keyword": "type",
      "message": "should be object",
      "params": Object {
        "type": "object",
      },
      "schemaPath": "#/definitions/OutputDescriptor/type",
    },
    Object {
      "dataPath": ".instruction.output[2].filename",
      "keyword": "type",
      "message": "should be string",
      "params": Object {
        "type": "string",
      },
      "schemaPath": "#/definitions/OutputDescriptor/properties/filename/type",
    },
    Object {
      "dataPath": ".instruction.output[2]",
      "keyword": "type",
      "message": "should be string",
      "params": Object {
        "type": "string",
      },
      "schemaPath": "#/properties/output/anyOf/1/items/anyOf/1/type",
    },
    Object {
      "dataPath": ".instruction.output[2]",
      "keyword": "anyOf",
      "message": "should match some schema in anyOf",
      "params": Object {},
      "schemaPath": "#/properties/output/anyOf/1/items/anyOf",
    },
    Object {
      "dataPath": ".instruction.output[3]",
      "keyword": "required",
      "message": "should have required property 'filename'",
      "params": Object {
        "missingProperty": "filename",
      },
      "schemaPath": "#/definitions/OutputDescriptor/required",
    },
    Object {
      "dataPath": ".instruction.output[3]",
      "keyword": "type",
      "message": "should be string",
      "params": Object {
        "type": "string",
      },
      "schemaPath": "#/properties/output/anyOf/1/items/anyOf/1/type",
    },
    Object {
      "dataPath": ".instruction.output[3]",
      "keyword": "anyOf",
      "message": "should match some schema in anyOf",
      "params": Object {},
      "schemaPath": "#/properties/output/anyOf/1/items/anyOf",
    },
    Object {
      "dataPath": ".instruction.output",
      "keyword": "type",
      "message": "should be string",
      "params": Object {
        "type": "string",
      },
      "schemaPath": "#/properties/output/anyOf/2/type",
    },
    Object {
      "dataPath": ".instruction.output",
      "keyword": "anyOf",
      "message": "should match some schema in anyOf",
      "params": Object {},
      "schemaPath": "#/properties/output/anyOf",
    },
    Object {
      "dataPath": ".instruction",
      "keyword": "anyOf",
      "message": "should match some schema in anyOf",
      "params": Object {},
      "schemaPath": "#/properties/instruction/anyOf",
    },
  ],
  "path": "multiple-errors/.schema.autogen.json",
  "value": undefined,
}
`;
